### DISCUSSION TITLE
How to integrate FreeRTOS ?

### DISCUSSION THREAD

**Discussion by fastlock** [earlephilhower/arduino-pico#519] [Q&A]
I'm having trouble integrating freeRTOS from the porting and example of https://github.com/yunkya2/pico-freertos-sample made for pico-SDK.
Any ideas?

**Comment by earlephilhower** [comment]
Moved this to a discussion since it’s not really a core issue.

You’re going to have to rewrite the whole core to build it under freertos.  It’s not impossible, but you can’t just include a header or source file to make it work.

**Reply by hfellner** [reply]
I was able to get the FreeRTOS SMP port for the RP2040 to run with your core. I did have to change a few lines in the startup code of the core, disabling its support to use the second CPU core, but since that functionality is provided by FreeRTOS it does not really pose a problem. Once I had done that it's enough to simply include a library to add FreeRTOS support to this core.

**Reply by earlephilhower** [reply]
Very nice.  Do you have any code to share?  Maybe it's something that could be yet another IDE option to make it nice and accessible.

I fear there's more to it than just adjusting the startup stuff.  For example, FreeRTOS is preemptive but we have places where a mutex is grabbed by either a HW IRQ or a class (i.e. in Serial* to make sure that we don't have both cores attempting to read the same byte and messing up the queues).  Feels like potential deadlocks or at least really bad performance in some cases.

The alternate-core-idle stuff, too, probably needs work (required for filesystem/EEPROM since you can't run out of flash while writing to it).

**Reply by hfellner** [reply]
I did make an Arduino library package using the FreeRTOS AVR library as a template, replacing the FreeRTOS files with the onces from the current RP2040 SMP branch port and uploaded it to a GitHub repository. I've uploaded it to GitHub [here](https://github.com/hfellner/RP2040_FreeRTOS).

I am aware of some other places that need changes, i.e. the LittleFS, EEPROM and Bootsel libraries, where the temporarily suspend the other core. This presents somewhat of a "chicken and egg" problem, as the core can not generally depend on libraries included later. I assume, this could be worked around using weak functions with a default implementation that works when FreeRTOS is not present, and an overridden implementation by the FreeRTOS library.

Another problem is the startup code in main.cpp, as the implementation of the RP2040 class performs static initialization that, if present, causes FreeRTOS to crash on initialization. So just the presence of the rp2040 object in main.cpp is enogh to prevent FreeRTOS from working. Also, since the FreeRTOS library (just like every other library adding FreeRTOS support for Arduino) works by implementing a initVariant() function that doesn't return, processing USB and serial events has to be done in the idle task in FreeRTOS. However, this creates an unwanted dependency between the library and the core.

Both of these problems, I guess, could be worked around by placing the code that does the RP2040 class initialization and the event processing in methods that then can be called from the appriate place - or not at all, in case of the RP2040 class. While this keeps a certain dependency between the core and the library, it is reduced to a defined interface instead of the library containing actual code from the core.

Concerning the Serial code, I would view this like most other situations where multiple cores or even multiple tasks/threads are present - using the same resource from multiple threads or even cores is inherently unsafe and requires external synchronisation.

**Reply by hfellner** [reply]
To illustrate what I meant I made a [fork of arduino-pico](https://github.com/hfellner/arduino-pico) and changed the startup code in main.cpp so that when FreeRTOS is present, the rp2040 object will not be initialized, and changed the FreeRTOS library to not use initVariant() but a separete function initFreeRTOS(), which is called from main() by the core after setup is complete. I moved the event processing to a separate function __loop(), which can be called from the FreeRTOS idle task. This way, when FreeRTOS is not included, the core behaves like always, but if it is, the FreeRTOS task scheduler is started after initialization, instead of the loop() function being called directly.

I'm testing a solution to the problem with suspending and resuming the other core, but I haven't committed that yet.

**Reply by hfellner** [reply]
So I've come to the conclusion that providing a working alternative to the RP2040::idleOtherCore() method while FreeRTOS is running requires support by FreeRTOS itself, such as a vTaskSuspendCore(uCore) API function that sends the given core into an idle loop not in flash. However, FreeRTOS's does not support functionality like this at the moment. I suspect it would be possible to implement such a function, but I have strong reservations against modifying the FreeRTOS kernel itself, as it makes integrating upstream updates difficult and potentially destructive.

It is theoretically possible to solve the problem using present FreeRTOS API functionality, but it would mean imposing limits about how tasks are set up (i.e. priorities), and I'd rather not do that, as it would probably prove a support nightmare.

So, for now, I would leave it up to the user. The only real problem seems to be querying the BOOTSEL button, because it will immediately crash any code running from Flash memory. Writing to EEPROM or Flash filesystem while accessing it simultaneously from another thread seems a complete NO NO to me, anyway, so I do not think it is much of a problem. I would argue that working with FreeRTOS requires enough knowledge to assume one knows about the dangers of accessing resources without synchronization.

I have updated the FreeRTOS library with a script that automates updating the FreeRTOS code from a working copy of the FreeRTOS smp branch and made it a compile option of the core accessible from the menu. If FreeRTOS is enabled, the then unsupported methods in the RP2040 class will be omitted from compilation, preventing problems at runtime if somebody erroneously calls them.

@earlephilhower
I kept the changes made to the arduino-pico core limited to as few as possible, with the hope it might be possible to integrate them into the core officially at some point in the future.

**Reply by earlephilhower** [reply]
Very nice!  I haven't had a chance to look at the code, but hopefully will get to it this weekend.  The no-filesystem/no-eeprom limitation seems a bit rough as you don't have any way to persist anything between powercycles in that case. :(  I'll look at how it's handling the SMP messaging and see if something with minimal changes can be added in...

**Reply by hfellner** [reply]
You absolutely can use LittleFS and EEPROM in FreeRTOS code, the library just doesn't provide synchronization in this case, that has to be done in user code, which can easily be done using FreeRTOS mutexes. I would discourage attempting to solve this from the library/core, as FreeRTOS is a realtime operating system, and as such must be able to make guarantees about the maximum time it takes from an event occurring to the point in time the corresponding handler task is executed. If there is implicit deactivation of a core by a library function and this is a task pinned to this core, this breaks the realtime characteristics of the OS.

**Reply by earlephilhower** [reply]
It's not synchronization so much as HW limitations.  You need to pull the other core into executing from RAM while you are writing/erasing the flash or the machine will crash since it can't read instructions while the flash is busy.

I'm not worried about two users trying to update LittleFS at the same time (but now that I think about it, this is probably also not allowed but can be handled w/mutexes like you suggested)...

**Reply by hfellner** [reply]
I see. I re-read the SDK flash API documentation, and found the warning about XIP access during flash programming, hidden in some generic paragraph. I agree it might be nice to be able to use LittleFS and EEPROM when FreeRTOS is just used because it's nice to have an OS with preemptive multitasking but realtime operation is not required.  
If or while realtime processing is required, IMHO these remain an absolute no-go.

**Reply by earlephilhower** [reply]
I've spent some time this AM and taken your repo and done some refactoring for #533.  Instead of copying files manually from upstream, the fork of FreeRTOS_RP2040 actually includes upstream as a submodule and uses `#include`s to make things work (the same way I keep LilttleFS upstream untouched for the core).  There were a few changes to your core repo patches needed to get things working, but nothing major.  Right now it builds and runs Blink.ino w/FreeRTOS fine but with many warnings (I always run w/`-wall -wextra`).

If you'd rather do the changes to this repo, let me know and I'll close #533 and send in a PR for your FreeRTOS_RP2040 repo for the changes I've done.  I used forks of upstream so I can add the special sauce needed to freeze the other core in RAM and whatever else pops up...

**Reply by hfellner** [reply]
Your approach with using proxy files that use `#include`s  is undoubtedly better, and I agree that using a fork of the original FreeRTOS kernel allows for local modifications better than copying the files. It does appear modifications to FreeRTOS are the only option to achieve EEPROM and flash writing support, as the FreeRTOS port uses the FIFO queue to force context switches on the other core for scheduling purposes and always ignores the queue data in its interrupt handler.

As for #533, just go ahead, as I fear I won't have much time to work in this in the immediate future, as unfortunately today is the last day of my vaction time.

**Reply by earlephilhower** [reply]
@hfellner I know you're back to work now, but the core has FreeRTOS SMP included now (with support for EEPROM/etc. flash writes).  If you do get a chance to mess with it and find something, do please let me know!

One gotcha I found was that under certain optimization settings, the CPU FIFO IRQ handler (the one that's responsible for actually doing task switching) would cause faults.  This was even happening under the untouched upstream version, and I think it is/was related to GCC 10's mangling of return instructions (i.e. GCC 10 will make the chip crash if you don't return a value from a function that's supposed to return a value whereas earlier versions would just return a random value and continue running).

My thinking is that GCC 10 is optimizing the return path code, but its optimizations aren't valid in the FIFO IRQ where you aren't returning to the original caller...you're swapping to a new stack/etc.  (The RPi team is using an earlier GCC AIUI so they wouldn't be seeing this...)

In any case by fixing the optimization level around that one IRQ handler it seems to be running for what I've thrown at it so far.

**Reply by earlephilhower** [reply]
I take it back.  FreeRTOS SMP seems to still be flaky.  The example fails now with it not actually creating the USB or CORE0 tasks, but adding another task in the code makes it work again.  Oy vey.

**Reply by hfellner** [reply]
@earlephilhower I followed the FreeRTOS branch (#533 ) over the previous weeks and it seems you put quite a lot of work into that, many thanks!
I check out the new version and ran some tests with it yesterday, i.e. creating 8 tasks (not pegged to a core) that just check the boot selector button and then sleep in a loop. When the tasks sleep for 10ms or more, everything runs fine, even for a whole day. If the tasks sleep for less than 10ms, the microcontroller hang after a few seconds. This occurs regardless of the optimization level. It only happens, when I actually check the boot selector button in the loop, but I want to run some further tests on that subject.

I am not a fan, however, of the automatic core0 and core1 tasks to achieve the traditional multicore support under FreeRTOS and that there's no way to disable that behavior, as when I am using an RTOS I usually want full and exact control over what happens and when. Also, this disabled support for the idle task. A real problem for me is that `setup()` is called from within the core0 task, after the FreeRTOS scheduler has already been started. This makes it impossible to perform some kind of check in user code during the boot process, and then NOT start the RTOS under some conditions, e.g. to perform a flash upgrade, or some other disruptive operation that require the system not to be running.

I really think that the core0 and core1 tasks should be optional, and I also think that there should be some support for the idle task. Maybe through making `loop` and `loop1` both weak symbols and only creating the corresponding tasks if they are present. And, maybe another weak symbol for a `void idle(void)` function, that gets called from the idle task. And while I think it's acceptable for `setup1()` to run from within the core1 task, I would greatly prefer `setup()` to be called before `__startFreeRTOS()`.

Also, I found that I had to add an `#include` for the FreeRTOS kernel's task.h to the `FreeRTOS.h` proxy file in the library, because otherwise it would be impossible to actually use FreeRTOS APIs, as they are not included. This may even be necessary for more FreeRTOS header files, but I haven't tested that yet.

**Reply by earlephilhower** [reply]
For the bootsel crash, can you give #572 a try?   I am pretty sure I've had the IRQ/idle order wrong since the beginning, but MT w/FreeRTOS would exacerbate the issue.

About the `#include <task.h>` need, isn't that standard to use FreeRTOS?  We haven't touched the base includes at all.

As for the setup/loop and setup1/loop, I am not so sure there's a logical way to get where you're going.  It doesn't make any sense to start `setup1` only after `setup` has completed...that's the point of the simple MT way.  Same argument for `loop1` waiting until `setup` has completed.

AIUI there is only one ApplicationIdleTask so I'm really not sure how you'd move both loops into it.  If you just moved `loop` in, it would run at the lowest prio while `loop1` would be above it and executed much more frequently.  And `loop1` would be a task and able to use `vTaskDelay` for `delay()` while `loop` would crash the system if it used that call (ask me how I know... :cry:)

Maybe a weak global could be introduced, say `__enablePureFreeRTOSMode` which you could override in your sketch to give you the very un-Arduino config you need?

**Reply by hfellner** [reply]
I tried the change from #572 but it didn't really change anything. If I have the tasks sleep for 1 ms, the controller hangs at the first cycle, and if I have them sleep for 4 ms, it runs seemingly stable until I whenever actually press the boot selector button and hangs then. At 10 ms, everthing runs stable.

About `#include <task.h>`, you're right, when using standard FreeRTOS one is expected to include the header explicitly some way like `#include <FreeRTOS/include/task.h>`, but with an Arduino library all the files are at the first level, but they are named far too ambiguously for that to be sensible, which probably is why the AVR FreeRTOS library restricts the include to the Arduino_FreeRTOS.h file, which includes all the necessary header files.

As for setup/loop, I see your point about not wanting to run `setup1` after `setup` has completed. Maybe we can find some compromise by adding a weak symbol `setup0` that, if present, will be run before the RTOS scheduler is started and that can be used for initialization that absolutely must be run before anything else. I would still very much like to make `setup` and `loop` weak symbols, too, and not create the `CORE0` task, like you already do with `setup1` and `loop1`, as that task is just wasteful, if `loop` is not used.

About the idle task I think you misunderstood me.. I did not mean for `loop` and `loop1` both to be called in the idle task, if it's used for that, `loop` would be enough, and there wouldn't need to be a `loop1`. But since you used them to provide multicore support, I had suggested something like an `idle` function, that could be called from the idle task.

As for this being a very un-Arduino config... this is just how any of the FreeRTOS libraries for Arduino work, most of them even explicitly state you're not supposed to call delay(), yield() or anything that might somehow block, and you're not supposed to use `loop` at all (unless they provide idle task support through it). If you're using an OS, with any microcontroller, and even any computer, you're always supposed to work the OS specific way. I think it's good you're providing a simple, comfortable, Arduino-like way to use FreeRTOS, but it is my opinion that people, who actually want FreeRTOS for its actual multitasking and/or realtime capabilities would be little interested in using that, as it takes away control over stack size and task priority from them, and just wastes the memory for the task and its stack, if unused. So, by making `setup`/`loop` weak symbols, everything will just work, if used the Arduino way, but it won't get into the way or waste resources if you actually want to work the FreeRTOS way.

**Reply by earlephilhower** [reply]
Wow, I guess my idea of Arduino-like and what the other ports are doing are way out of step!  It just seemed very natural to have a task per core setup/loop (and one for the USB handler).

Do you think you'd be able to rough out a PR to make it the way you know other Arduino FreeRTOS ports do it?  Or, do you have any SMP FreeRTOS Arduino ports I should look at?  I think the ESP32 has it's own port and doesn't behave like you're saying (since they need to have WIFI and peripheral tasks), so that's probably not a good example.

Even if not, the includes I can see a simple way to fix it.  Now I have `task.h` (and the other headers) redirected in the `src/` dir.  It would be a simple thing to remove them and then add the proper, `#include ../lib/FreeRTOS-Kernel/...` to the base `FreeRTOS.h`.  Let me give that a go as a simple start.

For the crash, can you post some sample code or do some investigation as to what's going in via GDB?  I spent far too much time in the guts of FreeRTOS fixing my mistakes (which caused random memory corruption, the most fun kind of memory corruption) so have some idea round the internals now.

**Reply by hfellner** [reply]
But if you just have one task per core (and furthermore, with the task pegged to the cores), then you don't actually have any advantage from using an RTOS, you could just as well use this Arduino core without FreeRTOS. Using a preemptive multitasking OS allows you to use one task for each function the controller is performing, and solves the problem associated with a central `loop` function, where everything has to be broken up in state machines, to allow for other actions to be performed while waiting on things to happen. This isn't necessary if each task is dedicated to a specific function the controller needs to perform, and a task can just simply wait/block, while other tasks are still able to run, contributing to less error prone and more easily readable code.

About the way other cores (or rather, libraries) are implementing FreeRTOS for Arduino, I did not mean other SMP cores, but rather FreeRTOS libraries for other cores in general - to my knowledge, there is only one other microcontroller with more than one core and FreeRTOS support for Arduino - ESP32 - but that is a bad comparison, since FreeRTOS is not optional on ESP32 - it always runs on top of it. Also, FreeRTOS for the ESP32 is not based on FreeRTOS SMP, but rather a custom implementation that predates the FreeRTOS SMP branch, and with a different API for SMP specific functionality. (But even the ESP32 specific FreeRTOS example states you're supposed to use tasks, and not the loop function.)

Regarding making `setup`/`loop` weak symbols, I am leaning towards a compile option like "Include CORE0/CORE1 default tasks" (enabled by default) rather than making `setup`/`loop` weak symbols, since that would lead to successful compilation of an Arduino sketch without a `setup` or `loop` function even when not using FreeRTOS, which for a typical Arduino sketch, is a clear error. What do you think?

I've tried different approaches regarding the `#include` for the FreeRTOS files, but it seems that FreeRTOS itself is meant to be compiled to a library, rather than included by source, as the implementation files misuse `#include <...>` and `#include "..."` statements, leading to compilation errors when the header files are not available at the first level. But I've found out I was under the wrong impression that the `Includes`-directive of the library.properties file allows only that particular file to be included - which is not the case, it just restricts the library to be included in the sketch only via this specific file, and that is exactly what is probably best in this case, as it allows for a subsequent `#include <task.h>`, but  prevents FreeRTOS from being included in the sketch, if some other library also has an include file "task.h". So I'd recommend keeping the library as it is, and just re-add the `Includes=FreeRTOS.h` directive in library.properties.

I can prepare a PR with the changes I described, that would default to the most Arduino-like way for an easy start with FreeRTOS, but with an option for more fine-grained control.

As for the crash.. that it is dependent on whether the boot selector button is actually pressed in certain situation has led me to believe that the problem is actually an electrical one, rather than a pure software problem, and that it might have something to do with a residual charge or signal slope being too slow on an enable pin (or something like that) after access to the flash memory is restored. So I added a short delay in Bootsel.cpp after re-enabling flash memory access, but before the other core and interrupts are resumed, and this seems to have solved the problem. I have 8 tasks constantly checking the boot selector button and then just sleeping for 0 milliseconds, i.e. switching to the next task without waiting. While this led to an immediate crash in every single case before, this way it has been running stable for almost a whole day, so far - regardless, how many times or how long I press the boot selector button.

As for the length of the delay - my tests have shown that 100 µs is too short, and still causes crashes/hangs sometimes, but 200 µs is enough for it to run really stable. However, if this really is the cause for the problem, this might mean that different RP2040 boards might behave differently in regards to this problem, as different capacitor/resistor values might change the required delay time. My tests were done with a standard Raspberry Pi Pico board.

Lastly - I'm really sorry about the long post.

**Reply by earlephilhower** [reply]
>Lastly - I'm really sorry about the long post.

Thanks for taking the time to make the core better, no worries!  👍

I get your discussion except for one point:  the CORE0 and CORE1 tasks now are at prio 4.  So, higher prio tasks will always get scheduled ahead of it of course.  And on any delay() lower priority tasks can be called (delay == vTaskDelay).  It seems to me with this kind of setup you can still get the hard realtime behavior by using prio >= 5.  (Note that USB is IRQ based on a 1ms increment so in FreeRTOS mode it's a prio-7 task to mimic that....which may be trouble if you're doing somethign depending on us-level accuracy on core0...)

A simple check shows I was wrong about renaming includes.  It's not so easy and would require patching FreeRTOS.  It's definitely doable but not as simple as `git mv`.

BOOTSEL is kind of a HW/SW hack the Pico guys came up with.  It pauses all other core0 and core1 operations while executing (which takes a non-trivial time IIRC), so honestly I think you'd not be able to use in any critical FreeRTOS app like you were discussing.  So it might not be worth looking into too hard.

**Comment by ogatatsu** [comment]
Hello
The SMP branch of FreeRTOS has an option to make sem/mutex/queue and sleep of Pico SDK work with FreeRTOS tasks

`configSUPPORT_PICO_SYNC_INTEROP`
`configSUPPORT_PICO_TIME_INTEROP`

https://github.com/earlephilhower/FreeRTOS-Kernel/blob/f3612388755827dbc3ac48f40764045f4897986f/portable/ThirdParty/GCC/RP2040/include/rp2040_config.h#L47-L64

but it seems that you need to include `freertos_sdk_config.h` in your Pico SDK build for this to work.

https://github.com/earlephilhower/FreeRTOS-Kernel/blob/f3612388755827dbc3ac48f40764045f4897986f/portable/ThirdParty/GCC/RP2040/CMakeLists.txt#L27
